libcxxabi/include/cxxabi.h

```diff
@@ -36,13 +36,19 @@ class type_info; // forward declaration
 
 // runtime routines use C calling conventions, but are in __cxxabiv1 namespace
 namespace __cxxabiv1 {
+
+struct __cxa_exception;
+
 extern "C"  {
 
 // 2.4.2 Allocating the Exception Object
 extern _LIBCXXABI_FUNC_VIS void *
 __cxa_allocate_exception(size_t thrown_size) throw();
 extern _LIBCXXABI_FUNC_VIS void
 __cxa_free_exception(void *thrown_exception) throw();
+// This function is an llvm extension
+extern _LIBCXXABI_FUNC_VIS __cxa_exception*
+__cxa_init_primary_exception(void* object, std::type_info* tinfo, void(_LIBCXXABI_DTOR_FUNC* dest)(void*)) throw();
```

Comment:
This probably requires availability annotations. CC @ldionne

Reply:
Actually, I would like to have the opinion of some other folks on this first. We generally try to keep libc++abi close to the specification and I'm really unsure about adding an "LLVM extension".

@rjmccall @MaskRay Do you have an opinion about this patch? I am a bit wary of adding extensions to the Itanium C++ ABI in libc++abi.

Reply:
To add some context to the discussion:

- gcc has this function as "GNU extension": https://github.com/gcc-mirror/gcc/blob/master/libstdc%2B%2B-v3/libsupc%2B%2B/cxxabi_init_exception.h#L70, which was added with the exact same performance reasoning in mind: https://gcc.gnu.org/legacy-ml/libstdc++/2016-08/msg00036.html
- as of recently `libcxxrt` has this function as well: https://github.com/libcxxrt/libcxxrt/blob/master/src/cxxabi.h#L211

Reply:
Coordinating to avoid divergence/ABI incompatibility with GCC is important, so that we're implementing a pre-existing GCC ABI extension is very important to mention!

The context that this is an extension that GCC already implements, and that LLVM is now implementing compatibly, should to be mentioned in the PR description and in the comments for this function. (Presuming that our implementation _is_ ABI-compatible -- which it of course needs to be!)

Given that, IMO, it's fine to make this ABI extension.

Reply:
Yeah, I think that changes everything indeed.

Reply:
> * as of recently `libcxxrt` has this function as well: [https://github.com/libcxxrt/libcxxrt/blob/master/src/cxx](https://github.com/libcxxrt/libcxxrt/blob/master/src/cxxabi.h#L211)

It's worth noting that you added the libcxxrt extension.

Reply:
> It's worth noting that you added the libcxxrt extension.

Agreed, didn't mean to try to trick anyone.

The point i'm trying to make is that
a) both libsupc++ and libcxxrt found the extension reasonable
b) libcxx could make use of the extension already present in aforementioned implementation, and it wouldn't be unreasonable to add it into libcxxabi as well